python/strings_udf/cpp/include/cudf/strings/udf/udf_string.hpp

```diff
@@ -0,0 +1,532 @@
+/*
+ * Copyright (c) 2020-2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cudf/strings/string_view.hpp>
+
+#include <cuda_runtime.h>
+
+// This header contains all class and function declarations so that it
+// can be included in a .cpp file which only has declaration requirements
+// (i.e. sizeof, conditionally-comparable, explicit conversions, etc).
+// The definitions are coded in udf_string.cuh which is to be included
+// in .cu files that use this class in kernel calls.
+
+namespace cudf {
+namespace strings {
+namespace udf {
+
+/**
+ * @brief Device string class for use with user-defined functions
+ *
+ * This class manages a device buffer of UTF-8 encoded characters
+ * for string manipulation in a device kernel.
+ *
+ * It's methods and behavior are modelled after std::string but
+ * with special consideration for UTF-8 encoded strings and for
```

Comment:
I do not expect anything to change. It may be possible that `udf_string` would encapsulate a `cuda::std::string` in the future but I don't think that is worth noting here.

Reply:
As mentioned in the doxygen comment, the `udf_string` has specialized logic for UTF-8 encoded string data whereas  the `std::string` (and `cuda::std::string`) do/would not. This is a significant difference which helps with cudf and libcudf integration. Moving forward I expect the `udf_string` _may contain_ a `cudf::std::string` mainly to help manage the char bytes.

Reply:
Ah I see. That is unfortunate, I was hoping that we would eventually be able to get rid of this class in favor of a `cuda::std::string`, but that makes sense. Do you think that the fact that we definitely need additional features relative to `cuda::std::string` means that it would eventually be OK to move this functionality into libcudf? I believe that one of the main original objections to adding this to libcudf was that it would be too close to a class that would eventually live in `cuda::std`, but if that's not the case then maybe we have a stronger case.